File: haskell/task1-2-v67.hs
{-Розбити список на два списки відповідно до умови – “бути чи не бути факторіалом деякого числа”.-}
splitFactorials :: [Int] -> ([Int], [Int])
splitFactorials [] = ([], [])
splitFactorials (x:xs) =
  let (factorials, others) = splitFactorials xs
  in if isFactorial x 2
       then (x:factorials, others)
       else (factorials, x:others)
  where
    isFactorial :: Int -> Int -> Bool
    isFactorial 1 k = True
    isFactorial n k
      | n <= 0 = False
      | n < k = False
      | mod n k == 0 = isFactorial (div n k) (k + 1)
      | otherwise = False
main :: IO ()
main = do
    putStrLn "--- Test 1 ---"
    putStrLn "Input: [1, 2, 3, 4, 5, 6, 24, 25]"
    putStrLn "Expected: ([1, 2, 6, 24],[3, 4, 5, 25])"
    putStrLn "Result:"
    print (splitFactorials [1, 2, 3, 4, 5, 6, 24, 25])

    putStrLn "\n--- Test 2 ---"
    putStrLn "Input: [2, 6, 24, 120]"
    putStrLn "Expected: ([2, 6, 24, 120],[])"
    putStrLn "Result:"
    print (splitFactorials [2, 6, 24, 120])

    putStrLn "\n--- Test 3  ---"
    putStrLn "Input: [-5, 0, 4, 10]"
    putStrLn "Expected: ([],[-5, 0, 4, 10])"
    putStrLn "Result:"
    print (splitFactorials [-5, 0, 4, 10])

    putStrLn "\n--- Test 4 ---"
    putStrLn "Input: []"
    putStrLn "Expected: ([],[])"
    putStrLn "Result:"
    print (splitFactorials [])
{-
    putStrLn "Enter list of numbers:"
    input <- getLine
    
    let numbers = map read (words input) :: [Int]
    putStrLn "\nResult:"
    print (splitFactorials numbers)
-}

File: prolog/task1-2-v67.pl
/* Розбити список на два списки відповідно до умови: "бути чи не бути факторіалом деякого числа" */

split_factorials([], [], []).

split_factorials([X|Xs], Factorials, Others) :-
    ( is_factorial(X, 2) ->
        Factorials = [X|Fs],
        Others = Os
    ;
        Factorials = Fs,
        Others = [X|Os]
    ),
    split_factorials(Xs, Fs, Os).

is_factorial(1, _).
is_factorial(N, K) :-
    N > 1,
    N >= K,
    0 is N mod K,
    N1 is N // K,
    K1 is K + 1,
    is_factorial(N1, K1).

/*
?- split_factorials([1, 2, 3, 4, 5, 6, 24, 25], F, O).
F = [1, 2, 6, 24],
O = [3, 4, 5, 25].

?- split_factorials([2, 6, 24, 120], F, O).
F = [2, 6, 24, 120],
O = [].

?- split_factorials([-5, 0, 4, 10], F, O).
F = [],
O = [-5, 0, 4, 10].

?- split_factorials([], F, O).
F = [],
O = [].*/